infer/component.go

```diff
@@ -110,3 +110,78 @@ func (rc *derivedComponentController[R, I, O]) Construct(
 			return res, err
 		})
 }
+
+// ** Implementations for creating an [InferredComponent] using existing Pulumi component programs. **
+
+// ComponentFn describes the type signature of a Pulumi custom component resource that users create.
+type ComponentFn[I any, O pulumi.ComponentResource] func(
+	*pulumi.Context, string, I, ...pulumi.ResourceOption) (O, error)
+
+// ProgramComponent creates an [InferredComponent] using functions and types that a existing Pulumi component program
+// would have implemented. The required inputs are the inputs and outputs struct, and the function that creates
+// the component resource.
+// See: https://www.pulumi.com/docs/iac/concepts/resources/components/#authoring-a-new-component-resource.
+func ProgramComponent[I any, O pulumi.ComponentResource](fn ComponentFn[I, O]) InferredComponent {
+	return &derivedProgramComponentController[I, O]{
+		construct: fn,
+	}
+}
+
+// derivedProgramComponentController is a controller for a component resource authored as a Pulumi program.
+type derivedProgramComponentController[I any, O pulumi.ComponentResource] struct {
+	// construct is the function that works on the component resource.
+	construct ComponentFn[I, O]
+}
+
+func (rc *derivedProgramComponentController[I, O]) GetToken() (tokens.Type, error) {
+	return getToken[O](nil)
+}
+
+func (derivedProgramComponentController[I, O]) isInferredComponent() {}
```

Comment:
I'm curious what is the purpose of this method?

Reply:
This is to prevent `InferredComponent` from being implemented in another package. Since this is an unexported  method, the Go compiler will complain we try to implement this outside of the `infer` package.

Since there is no strong reason to enable this functionality, I'm leaving it as is for now.

Reply:
Ah I see it on the type definition above now. We should probably have a comment on the type definition explaining it's just there as a sentinel